### PR TITLE
Improve Venmic Usability

### PR DESCRIPTION
**Pull request by Curve**
- [x] Allow to bypass `pipewire-pulse` warning
   image
- [x] Prevent `New Audio Device Detected` Modal with venmic
   image
- [x] Hide venmic from `Input Devices` in Audio Settings

### CODE REVIEW

File: src/renderer/patches/hideSwitchDevice.tsx [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+            find: "lastOutputSystemDevice.justChanged",
+            replacement: {
+                match: /((\w)\.default\.getState\(\).neverShowModal)/,
+                replace: "$1 || $self.shouldIgnore($2)"
+            }
+        }
+    ],
+
+    shouldIgnore(state: any) {
+        return Object.keys(state?.default?.lastDeviceConnected)?.[0] === "vencord-screen-share";
```
Comment (Vendicated) [inline src/renderer/patches/hideSwitchDevice.tsx:21]: Object.keys(undefined) throws so this optional chaining makes little sense
```suggestion
        return Object.keys(state?.default?.lastDeviceConnected ?? {})?.[0] === "vencord-screen-share";
```

File: src/renderer/patches/hideSwitchDevice.tsx [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+ */
+
+import { addPatch } from "./shared";
+
+addPatch({
+    patches: [
+        {
+            find: "lastOutputSystemDevice.justChanged",
+            replacement: {
+                match: /((\w)\.default\.getState\(\).neverShowModal)/,
```
Comment (Vendicated) [inline src/renderer/patches/hideSwitchDevice.tsx:14]:
```suggestion
                match: /(\i)\.default\.getState\(\).neverShowModal/,
```

File: src/renderer/patches/hideSwitchDevice.tsx [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+                replace: "$1 || $self.shouldIgnore($2)"
```
Comment (Vendicated) [inline src/renderer/patches/hideSwitchDevice.tsx:15]:
```suggestion
                replace: "$& || $self.shouldIgnore($1)"
```

**Comment by Vendicated** [comment]
got some conflicts now with katlyns pr being merged

**Comment by Curve** [comment]
> got some conflicts now with katlyns pr being merged

Fixed :)

**Comment by Justsnoopy30** [comment]
This PR breaks stream audio for me on Linux. I compared the latest Vesktop commit to the latest Vesktop commit with this PR reverted.

I also checked that it was not caused by the Venmic version bump from 3.4.1 to 3.4.2. It's only broken with the actual code changes from the PR, and the logs look about the same (just to show, I'll put them both below).

Log before PR:
```[arRPC > ipc] listening at /run/user/1000/discord-ipc-0
[arRPC > websocket] listening on 6463
[arRPC > process] started
[arRPC > process] detected game! Geometry Dash
[2024-04-20 14:30:31.255] [venmic] [info] [patchbay] (get) running venmic 3.4.1
[2024-04-20 14:30:31.266] [venmic] [info] [patchbay] (handle) found default metadata: 35
[2024-04-20 14:30:31.266] [venmic] [info] [patchbay] (meta_update) speaker name: "alsa_output.usb-1532_Razer_BlackShark_V2_Pro_2.4_O001000007-00.analog-stereo"
[2024-04-20 14:30:34.359] [venmic] [info] [patchbay] (create_mic) created: 209
```

Log after: 
```[arRPC > ipc] listening at /run/user/1000/discord-ipc-0
[arRPC > websocket] listening on 6463
[arRPC > process] started
[arRPC > process] detected game! Geometry Dash
[2024-04-20 14:43:29.423] [venmic] [info] [patchbay] (get) running venmic 3.4.2
[2024-04-20 14:43:29.434] [venmic] [info] [patchbay] (handle) found default metadata: 35
[2024-04-20 14:43:29.435] [venmic] [info] [patchbay] (meta_update) speaker name: "alsa_output.usb-1532_Razer_BlackShark_V2_Pro_2.4_O001000007-00.analog-stereo"
[2024-04-20 14:43:31.465] [venmic] [info] [patchbay] (create_mic) created: 166
```

Output of `pactl info`: 
```Server String: /run/user/1000/pulse/native
Library Protocol Version: 35
Server Protocol Version: 35
Is Local: yes
Client Index: 8678
Tile Size: 65472
User Name: justsnoopy30
Host Name: main-computer
Server Name: PulseAudio (on PipeWire 0.3.79)
Server Version: 15.0.0
Default Sample Specification: float32le 2ch 48000Hz
Default Channel Map: front-left,front-right
Default Sink: alsa_output.usb-1532_Razer_BlackShark_V2_Pro_2.4_O001000007-00.analog-stereo
Default Source: alsa_input.usb-1532_Razer_BlackShark_V2_Pro_2.4_O001000007-00.mono-fallback
Cookie: 7ac8:82dc
```

**Comment by Curve** [comment]
> This PR breaks stream audio for me on Linux. I compared the latest Vesktop commit to the latest Vesktop commit with this PR reverted.
> 
> I also checked that it was not caused by the Venmic version bump from 3.4.1 to 3.4.2. It's only broken with the actual code changes from the PR, and the logs look about the same (just to show, I'll put them both below).
> 
> Log before PR:
> ```[arRPC > ipc] listening at /run/user/1000/discord-ipc-0
> [arRPC > websocket] listening on 6463
> [arRPC > process] started
> [arRPC > process] detected game! Geometry Dash
> [2024-04-20 14:30:31.255] [venmic] [info] [patchbay] (get) running venmic 3.4.1
> [2024-04-20 14:30:31.266] [venmic] [info] [patchbay] (handle) found default metadata: 35
> [2024-04-20 14:30:31.266] [venmic] [info] [patchbay] (meta_update) speaker name: "alsa_output.usb-1532_Razer_BlackShark_V2_Pro_2.4_O001000007-00.analog-stereo"
> [2024-04-20 14:30:34.359] [venmic] [info] [patchbay] (create_mic) created: 209
> ```
> 
> Log after: 
> ```[arRPC > ipc] listening at /run/user/1000/discord-ipc-0
> [arRPC > websocket] listening on 6463
> [arRPC > process] started
> [arRPC > process] detected game! Geometry Dash
> [2024-04-20 14:43:29.423] [venmic] [info] [patchbay] (get) running venmic 3.4.2
> [2024-04-20 14:43:29.434] [venmic] [info] [patchbay] (handle) found default metadata: 35
> [2024-04-20 14:43:29.435] [venmic] [info] [patchbay] (meta_update) speaker name: "alsa_output.usb-1532_Razer_BlackShark_V2_Pro_2.4_O001000007-00.analog-stereo"
> [2024-04-20 14:43:31.465] [venmic] [info] [patchbay] (create_mic) created: 166
> ```
> 
> Output of `pactl info`: 
> ```Server String: /run/user/1000/pulse/native
> Library Protocol Version: 35
> Server Protocol Version: 35
> Is Local: yes
> Client Index: 8678
> Tile Size: 65472
> User Name: justsnoopy30
> Host Name: main-computer
> Server Name: PulseAudio (on PipeWire 0.3.79)
> Server Version: 15.0.0
> Default Sample Specification: float32le 2ch 48000Hz
> Default Channel Map: front-left,front-right
> Default Sink: alsa_output.usb-1532_Razer_BlackShark_V2_Pro_2.4_O001000007-00.analog-stereo
> Default Source: alsa_input.usb-1532_Razer_BlackShark_V2_Pro_2.4_O001000007-00.mono-fallback
> Cookie: 7ac8:82dc
> ```

Can you also share the debug logs (see venmic readme) when an application is shared before and after the commit?

**Comment by Justsnoopy30** [comment]
> Can you also share the debug logs (see venmic readme) when an application is shared before and after the commit?

Sure. 

Before: 
[venmic.log](https://github.com/Vencord/Vesktop/files/15049673/venmic.log)

After: 
[venmic.log](https://github.com/Vencord/Vesktop/files/15049674/venmic.log)

**Comment by Curve** [comment]
> > Can you also share the debug logs (see venmic readme) when an application is shared before and after the commit?
> 
> Sure.
> 
> Before: [venmic.log](https://github.com/Vencord/Vesktop/files/15049673/venmic.log)
> 
> After: [venmic.log](https://github.com/Vencord/Vesktop/files/15049674/venmic.log)

Thanks! It's definitely an issue with the JS-Side, linking works properly. I'll take a look tonight, I probably accidentally broke the lookup of the Venmic node for the screen share

**Pull request merged**